spec/unit/pops/loaders/loader_paths_spec.rb

```diff
@@ -38,6 +38,7 @@ def foo; end;
     expect(effective_paths.size).to eq(1)
     expect(module_loader.path_index.size).to eq(1)
     path_index = module_loader.path_index
-    expect(path_index).to include(File.join(module_dir, 'lib', 'puppet', 'functions', 'foo4x.rb'))
+    expected = File.join(module_dir, 'lib', 'puppet', 'functions', 'foo4x.rb')
+    expect(path_index).to include(Dir.glob(expected).first)
```

Comment:
I did another spike last night on calling `File.expand_path` in a few key locations, so that paths were always of the desired format (which I think would make things simpler).  I used `dtrace` to make sure `expand_path` doesn't make any extra `fstat64` calls on OSX.

However, `expand_path` on Windows didn't have the desired behavior, as it only expanded the last segment of the path with a `~`:

```
[7] pry(#<Puppet::Module>)> File.expand_path('C:/Users/ADMINI~1/')
=> "C:/Users/Administrator"
[8] pry(#<Puppet::Module>)> File.expand_path('C:/Users/ADMINI~1/AppData')
=> "C:/Users/ADMINI~1/AppData"
```

One thought was building up a new `FileSystem` method called `expand_path_recursive` or similar that made sure we got the desired behavior.